### ISSUE TITLE
`subset` arg of H5PYDataset overrides `which_sets`

### ISSUE THREAD

**Issue by lamblin** [mila-udem/fuel#208]
The documentation clearly states that the subset is interpreted "within the context of the split", but:
```python
>>> from fuel.datasets import MNIST
>>> train = MNIST(which_sets=['train'], sources=['targets'])
>>> test = MNIST(which_sets=['test'], sources=['targets'])
>>> subtest = MNIST(which_sets=['test'], sources=['targets'], subset=slice(0, 10000))
>>> train.get_data(request=slice(0, 10))
(array([[5],
        [0],
        [4],
        [1],
        [9],
        [2],
        [1],
        [3],
        [1],
        [4]], dtype=uint8),)
>>> test.get_data(request=slice(0, 10))
(array([[7],
        [2],
        [1],
        [0],
        [4],
        [1],
        [4],
        [9],
        [5],
        [9]], dtype=uint8),)
>>> subtest.get_data(request=slice(0, 10))
(array([[5],
        [0],
        [4],
        [1],
        [9],
        [2],
        [1],
        [3],
        [1],
        [4]], dtype=uint8),)
```

**Comment by dwf** [comment]
Confirmed. Ergh.

**Comment by lamblin** [comment]
The following is working correctly, though:
```python
>>> subtest = MNIST(which_sets=['test'], sources=['targets'], subset=list(xrange(10000)))
```
@vdumoulin is having a look right now.

**Comment by vdumoulin** [comment]
The offending block of code is [the following](https://github.com/mila-udem/fuel/blob/master/fuel/datasets/hdf5.py#L528-L536). The logic is plain wrong.

Thinking of it, I'm wondering if we should simply not care about maintaining slices if the split is defined by a slice and the subset is also a slice. Always converting the split into a list of indices and using `subset` to slice/index through that would make the logic much simpler, both for `load` and `_out_of_memory_get_data`.

**Comment by dwf** [comment]
Also absurdly inefficient for large numbers of examples...

**Comment by lamblin** [comment]
Fixing the slice/slice logic would not be hard, as long as we are not planning support for negative steps. Is it OK to put that limitation?

**Comment by rizar** [comment]
Will this be fixed by #223 ? Seems like yes.

**Comment by vdumoulin** [comment]
I confirm that it's fixed by #223. Closing.
